Source file trees

Thyrum opened the pull request:
Implement source file tree support.

Thyrum committed:
Implement file tree copying using .SECONDEXPANSION

File tree support has been implemented using the
[.SECONDEXPANSION](https://www.gnu.org/software/make/manual/html_node/Secondary-Expansion.html)
directive and a new
[`rwildcard`](https://blog.jgc.org/2011/07/gnu-make-recursive-wildcard-function.html)
function for recursively finding source files.

One thing missing is adding the directory versions in `DEPDIR` and
`OBJDIR` to the `.PRECIOUS` directive. This should probably be added
before merging

Thyrum committed:
Add entire file tree to .PRECIOUS directive

Previously one could get an error like
```
build/obj/release/util/.make: unlink: build/obj/release/util/.: Is a directory
```
because make tries to clean up intermediate build targets. This commit
adds the directory targets to the PRECIOUS target removing these errors.

Thyrum commented:
Please be carefull when merging this since it might conflict with #4, that is to say, the code in this branch might need to be altered to also support multiple source file extensions.

Thyrum committed:
Reduce intrusiveness of make clean

Rework make clean as discussed in #3. Now we only delete our build files
and all empty subdirectories of $(OBJDIR), $(DEPDIR) and $(BINDIR).
We also delete all empty directories along the relative path of
$(OBJDIR), $(DEPDIR) and $(BINDIR).

Thyrum commented:
The clean commands don't work properly on this branch. The versions on #6 should work so it is probably wise to merge that branch into this one first.

Edit: since 3eec3c7, the clean command has been fixed.

avdstaaij committed:
Simplify obj and dep dir expansion in .PRECIOUS

Since all that pathsubst was doing here was adding a suffix, it could be
replaced by the simpler addsuffix.

avdstaaij committed:
Improve make clean error handling

Rewrote the find commands from make c to throw errors when the
directories to clean do not exist. This has two advantages:
- The 2>/dev/null redirection could be removed, allowing the user to see
  possibly useful messages about file removal failures again
- Make no longer reports ignored errors when the directories to clean do
  do not exist

The addition of "|| true" is again to silence make's "ignored error"
output.

Errors from the binary removal commands now also won't stop make,
consistent with the other removals. This may allow make to clean some
other files even if removing a binary fails.

avdstaaij committed:
Change make c to use $(RM)

This way, make c uses the user-specified deletion command instead of
letting find do the deletions

avdstaaij committed:
Change make c to remove obj and dep directories

If the object and dependency files are removed, their directories should
be removed as well.

avdstaaij commented on Makefile:
```diff
@@ -70,12 +70,18 @@ MKBIN_D = $(CXX) $(CXXFLAGS) $(FLAGS_D) $^ $(LDFLAGS) -o $@
 
 ################################################################################
 
+# Functions
+
+rwildcard=$(foreach d,$(wildcard $1*),$(call rwildcard,$d/,$2) $(filter $(subst *,%,$2),$d))
+
+################################################################################
+
```
I think adding a new "segment" for just this function is a little jarring, considering the segment below it is much larger and contains many different kinds of statements.

If we are going to add more segments (as is also done in #4), we may want to rewrite this line:
`# The lines above the #-line can be altered to change the build behavior.`

avdstaaij commented on Makefile:
```diff
@@ -121,23 +128,24 @@ $(BIN_D): $(OBJS_D) | $(BINDIR)/.
 
-$(OBJDIR_R)/%.o: $(SRCDIR)/%$(SRCEXT) $(DEPDIR_R)/%.d | $(OBJDIR_R)/. $(DEPDIR_R)/.
+.SECONDEXPANSION:
+$(OBJDIR_R)/%.o: $(SRCDIR)/%$(SRCEXT) $(DEPDIR_R)/%.d | $$(dir $$@). $$(dir $(DEPDIR_R)/$$*.d).
 	$(CXX) $(CXXFLAGS) $(FLAGS_R) $(INCFLAGS) -c $< -o $@ $(DEPGENFLAGS_R)
 	$(POSTCOMPILE_R)
 
-$(OBJDIR_D)/%.o: $(SRCDIR)/%$(SRCEXT) $(DEPDIR_D)/%.d | $(OBJDIR_D)/. $(DEPDIR_D)/.
+$(OBJDIR_D)/%.o: $(SRCDIR)/%$(SRCEXT) $(DEPDIR_D)/%.d | $$(dir $$@). $$(dir $(DEPDIR_D)/$$*.d).
 	$(CXX) $(CXXFLAGS) $(FLAGS_D) $(INCFLAGS) -c $< -o $@ $(DEPGENFLAGS_D)
 	$(POSTCOMPILE_D)
```
I think there should be a newline after `.SECONDEXPANSION`, since it affects more than just the rule immediately below.
We may even want to move `make c` and `make clean` to above `.SECONDEXPANSION`, since those rules don't need it. What is your opinion on this?

Thyrum replied:
I was also considering this, I just didn't take the time to put it in a commit yet. I do agree on moving `clean` and `c`.

avdstaaij reviewed (changes requested):
The recursive source directory support works perfectly!

The cleaning logic is not quite yet perfect (see my comment in #3), so I will remove the issue link for now. I have also changed it a bit: see my commits.

We may want to squash some commits before merging this. For example, fda48cbca57bf2aef5c13507f504a7303c5ff1ea was completely replaced by c8285a6e5ac5301601ef5e203e9da8d9ffe8a94d. (Edit: the hashes have changed after a rebase.) (FIXED)

The readme still needs to be amended before this is merged. (FIXED)

Thyrum commented:
> The cleaning logic is not quite yet perfect

It is true that cleaning still goes wonky when `BINDIR` is set to `.` but it wouldn't suffice to simply check if `BINDIR==.` since when building libraries it is not uncommon to set `BINDIR` to `..`. I honestly don't know how to handle this properly.

> The readme still needs to be amended before this is merged.

I don't think the `README` has to mention the adjusted cleaning (since it practically already claims it works this way). I also don't know if the `README` should mention the source file trees since this might be expected behavior. If you do have a good way of wording the added source file trees, please do!

Thyrum committed:
Move .SECONDEXPANSION directive to the bottom

Since .SECONDEXPANSION affects all following directives, we move this to
the bottom. This way, only the directives that require it are affected.

Thyrum committed:
Add/improve section headers

Also added a fancy banner.

avdstaaij committed:
Add source file tree support feature to README

avdstaaij committed:
Fix some small errors in the README

Removed some non-functional trailing whitespace and made the tagline
capitalization consistent with the repository description.

avdstaaij committed:
Remove publication date from Makefile

We were most certainly going to forget to update this date, so I
believed it best to just remove it.

avdstaaij reviewed (approved):
I have added the source file trees to the feature list of the README. I believe it to be special enough to warrant a mention. I have also made a couple more small changes to the README.

I have also cleaned up the history a little: I have squashed a commit that was replaced by the next and a commit that was only a fixup of a previous one, and I have moved some changes from "Reduce intrusiveness of make clean" to "Implement file tree copying using .SECONDEXPANSION".

I believe this branch is now ready to be integrated into master. I belive this feature to be more important than the one from #4, so it's probably best if this branch is integrated first and #4 is rebased on top of it.

@Thyrum, do you agree with my (history) changes, in particular the changes to your commits?

Thyrum commented:
I agree with the changes, looks good! Time to merge.

Thyrum commented:
Has been merged manually
